text/0121-data-types.md

```diff
@@ -0,0 +1,435 @@
+But, the distance between `Bruce` and `Vin` is equal to `8`. The same goes for `Bruce` and `Diesel`, or `Willis` and `Vin`, or also `Willis` and `Diesel`.
+
+Let's see another example. Given two documents:
+
+```json
+[
+    {
+        "movie_id": "001",
+        "description": "Bruce.Willis"
+    },
+    {
+        "movie_id": "002",
+        "description": "Bruce super Willis"
+    }
+]
+```
+
+When making a query on `Bruce Willis`, `002` will be the first document returned and `001` will be the second one.
+This will happen because the proximity distance between `Bruce` and `Willis` is equal to `2` in the document `002` whereas the distance between `Bruce` and `Willis` is equal to `8` in the document `001` since the full stop is a hard space.
+
+### 3.2. Array Tokenization
+
+An array is recursively broken into separate string tokens, which means separate words. After the tokenizing process, each word is indexed and stored in the global dictionary of the corresponding index.
+
+#### 3.2.1. Examples
+
+The following input:
+
+```json
+[
+    [
+        "Bruce Willis",
+        "Vin Diesel"
+    ],
+    "Kung Fu Panda"
+]
+```
+
+Will be processed as if all elements were arranged at the same level:
+
+```json
+"Bruce Willis. Vin Diesel. Kung Fu Panda."
+```
+
+The strings above will be separated by soft and hard spaces exactly as explained in the string example.
+
+> See [3.1.1. Examples section](#311-examples).
+
+### 3.3. Object
+
+#### 3.3.1. Example
+
+The following input:
+
+```json
+{
+    "movie_id": "1564saqw12ss",
+    "title": "Kung fu Panda"
+}
+```
+
+In the example above, `movie_id`, `1564saqw12ss`, `title`, `Kung fu Panda` are all considered sentences. The colon `:` and comma `,` characters are used as separators.
+
+`"movie_id. 1564saqw12ss. title. Kung fu Panda."`
+
+These sentences will be separated by soft and hard spaces precisely as explained in the string example.
+
+### 3.4. Nested Structures
+
+Nested structures (e.g. `Object`, `Array of Objects`, etc) are internally flattened to a document's root level.
+
+It allows expressing a nested field in all Meilisearch parameters that accept document attributes.
+
+Meilisearch accepts the `.` (dot-notation) to express a nested field location in a document structure.
+
+### 3.4.1. Examples
+
+#### 3.4.1.1. Object
+
+The following JSON document:
+
+```json
+    {
+        "a": {
+            "b": "c",
+            "d": "e",
+            "f": "g"
+        }
+    }
+```
+
+Flattens to:
+
+```json
+{
+    "a.b": "c",
+    "a.d": "e",
+    "a.f": "g"
+}
+```
+
+#### 3.4.1.2. Array of objects
+
+The following JSON document:
+
+```json
+{
+    "a": [
+        { "b": "c" },
+        { "b": "d" },
+        { "b": "e" },
+    ]
+}
+```
+
+Flattens to:
+
+```json
+{
+    "a.b": ["c", "d", "e"],
+}
+```
+
+#### 3.4.1.3. Array of objects mixed with scalar value
```

Comment:
Yes! This is it

Reply:
I don't know if it's a good idea but it would be fairly easy for us (=the core team) to provide a wasm front end json flatener so people could test it and maybe understand better what is going on.
But doing the front end thingy might take too much time idk